src/effects/effectslot.cpp

```diff
@@ -68,154 +80,301 @@ EffectSlot::EffectSlot(const QString& group,
+    VERIFY_OR_DEBUG_ASSERT(m_pManifest != nullptr) {
+        return;
+    }
 
-EffectButtonParameterSlotPointer EffectSlot::addEffectButtonParameterSlot() {
-    EffectButtonParameterSlotPointer pParameter = EffectButtonParameterSlotPointer(
-            new EffectButtonParameterSlot(m_group, m_buttonParameters.size()));
-    m_buttonParameters.append(pParameter);
-    m_pControlNumButtonParameterSlots->forceSet(
-            m_pControlNumButtonParameterSlots->get() + 1);
-    return pParameter;
+    VERIFY_OR_DEBUG_ASSERT(m_pEngineEffect == nullptr) {
+        return;
+    }
+
+    m_pEngineEffect = new EngineEffect(m_pManifest,
+            activeInputChannels,
+            m_pEffectsManager,
+            std::move(pProcessor));
+
+    EffectsRequest* request = new EffectsRequest();
+    request->type = EffectsRequest::ADD_EFFECT_TO_CHAIN;
+    request->pTargetChain = m_pEngineEffectChain;
+    request->AddEffectToChain.pEffect = m_pEngineEffect;
+    request->AddEffectToChain.iIndex = m_iEffectNumber;
+    m_pEffectsManager->writeRequest(request);
 }
 
-EffectPointer EffectSlot::getEffect() const {
-    return m_pEffect;
+void EffectSlot::removeFromEngine() {
+    VERIFY_OR_DEBUG_ASSERT(isLoaded()) {
+        return;
+    }
+
+    EffectsRequest* request = new EffectsRequest();
+    request->type = EffectsRequest::REMOVE_EFFECT_FROM_CHAIN;
+    request->pTargetChain = m_pEngineEffectChain;
+    request->RemoveEffectFromChain.pEffect = m_pEngineEffect;
+    request->RemoveEffectFromChain.iIndex = m_iEffectNumber;
+    m_pEffectsManager->writeRequest(request);
+
+    m_pEngineEffect = nullptr;
 }
 
-unsigned int EffectSlot::numParameterSlots() const {
-    return m_parameters.size();
+void EffectSlot::updateEngineState() {
+    if (!m_pEngineEffect) {
+        return;
+    }
+    sendParameterUpdate();
+    for (auto const& pParameter : m_parameters) {
+        pParameter->updateEngineState();
+    }
 }
 
-unsigned int EffectSlot::numButtonParameterSlots() const {
-    return m_buttonParameters.size();
+void EffectSlot::sendParameterUpdate() {
+    if (!m_pEngineEffect) {
+        return;
+    }
+    EffectsRequest* pRequest = new EffectsRequest();
+    pRequest->type = EffectsRequest::SET_EFFECT_PARAMETERS;
+    pRequest->pTargetEffect = m_pEngineEffect;
+    pRequest->SetEffectParameters.enabled = m_pControlEnabled->get();
+    m_pEffectsManager->writeRequest(pRequest);
 }
 
-void EffectSlot::slotEnabled(double v) {
-    //qDebug() << debugString() << "slotEnabled" << v;
-    if (m_pEffect) {
-        m_pEffect->setEnabled(v > 0);
+EffectState* EffectSlot::createState(const mixxx::EngineParameters& bufferParameters) {
+    VERIFY_OR_DEBUG_ASSERT(m_pEngineEffect != nullptr) {
+        return new EffectState(bufferParameters);
     }
+    return m_pEngineEffect->createState(bufferParameters);
 }
 
-void EffectSlot::slotEffectEnabledChanged(bool enabled) {
-    m_pControlEnabled->set(enabled);
+EngineEffect* EffectSlot::getEngineEffect() {
```

Comment:
This function is being used in EffectParameter::updateEngineState, I've not removed it yet.

Reply:
I looked through EffectParameter and this is the only reason it needs a pointer to EffectSlot. I think it would be better to pass the EffectParameter constructor a pointer to the EngineEffect than EffectSlot, then this function could be removed.